clang/test/Preprocessor/macro-reserved-attrs.cpp

```diff
@@ -0,0 +1,33 @@
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cxx11 -pedantic -std=c++11 %s
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cxx14 -pedantic -std=c++14 %s
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cxx17 -pedantic -std=c++17 %s
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cxx20 -pedantic -std=c++20 %s
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cxx23 -pedantic -std=c++23 %s
+// RUN: %clang_cc1 -fsyntax-only -verify=expected,cxx26 -pedantic -std=c++26 %s
+
+#define noreturn 1           // cxx11-warning {{keyword is hidden by macro definition}}
```

Comment:
Please also add test cases for `#undef` of the name when it's not currently defined as a macro, as that's also undefined behavior. (@ldionne @jwakely any interest in an LWG issue over this? It's harmless to do `#undef noreturn` or whatever because these identifiers aren't macros to begin with; and if they were, the user hit the UB already.)

Reply:
I think that rule is there so that implementations can do:

`#define noreturn _Some_unspecified_noreturn_feature`

Reply:
How would that work? `int noreturn = 1;` is valid code AFAICT.

Reply:
What benefit would there be in allowing users to undef them? Just reducing the amount of UB it's possible to have?

Reply:
> How would that work? `int noreturn = 1;` is valid code AFAICT.

Yes, for attributes, that's true. For keywords it's not though.

The rule was changed between C++03 and C++11 to include #undef, I don't know why though.

I still don't see any advantage in allowing users to `#undef noreturn`.

Reply:
Yes, EB seems like a good choice here. It could be erroneously ignored, so if it's not diagnosed it has to be harmless.